### PR TITLE
Implementation of get_used_cells_and_allocated_size

### PR DESCRIPTION
**Pull request by Jrigada**
# Implementation of get_used_cells_and_allocated_size

## Description

Description of the pull request changes and motivation.

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

**Review by fmoletta — APPROVED** [review]

File: src/vm/runners/builtin_runner/bitwise.rs [hunk 1]
```diff
@@ -115,6 +119,32 @@ impl BitwiseBuiltinRunner {
...
+                .to_usize()
+                .ok_or(MemoryError::InsufficientAllocatedCells)?;
+        if vm.current_step < min_step {
+            Err(MemoryError::InsufficientAllocatedCells)
+        } else {
+            let builtin = BuiltinRunner::Bitwise(self);
+            let used = builtin.get_used_cells(vm)?;
+            let size = cells_per_instance
+                * safe_div(&bigint!(vm.current_step), &bigint!(ratio))
+                    .map_err(|_| MemoryError::InsufficientAllocatedCells)?;
```
Comment (fmoletta) [inline src/vm/runners/builtin_runner/bitwise.rs:143]: There is no need for this value to be a BigInt

File: src/vm/runners/builtin_runner/bitwise.rs [hunk 2]
```diff
@@ -115,6 +119,32 @@ impl BitwiseBuiltinRunner {
...
+        let ratio = self
+            ._ratio
+            .to_usize()
+            .ok_or(MemoryError::InsufficientAllocatedCells)?;
+        let cells_per_instance = self.cells_per_instance;
+        let min_step = ratio
+            * self
+                .instances_per_component
+                .to_usize()
+                .ok_or(MemoryError::InsufficientAllocatedCells)?;
```
Comment (Oppen) [inline src/vm/runners/builtin_runner/bitwise.rs:135]: `.instances_per_component` is an `u32`, just use `as usize`.

File: src/vm/runners/builtin_runner/bitwise.rs [hunk 3]
```diff
@@ -115,6 +119,32 @@ impl BitwiseBuiltinRunner {
...
     }
 
+    pub fn get_used_cells_and_allocated_size(
+        self,
+        vm: &VirtualMachine,
+    ) -> Result<(usize, BigInt), MemoryError> {
+        let ratio = self
+            ._ratio
+            .to_usize()
+            .ok_or(MemoryError::InsufficientAllocatedCells)?;
```
Comment (Oppen) [inline src/vm/runners/builtin_runner/bitwise.rs:129]: Same here.

File: src/vm/runners/builtin_runner/ec_op.rs [hunk 4]
```diff
@@ -198,6 +202,36 @@ impl EcOpBuiltinRunner {
...
+        println!("asd");
+        let cells_per_instance = self.cells_per_instance;
+        let min_step = ratio
+            * self
+                .instances_per_component
+                .to_usize()
+                .ok_or(MemoryError::InsufficientAllocatedCells)?;
+        println!("asd");
+        if vm.current_step < min_step {
+            println!("asd");
```
Comment (Oppen) [inline src/vm/runners/builtin_runner/ec_op.rs:224]:
```suggestion
        let ratio = self._ratio as usize;
        let cells_per_instance = self.cells_per_instance;
        let min_step = ratio * self.instances_per_component as usize;
        if vm.current_step < min_step {
```

File: src/vm/runners/builtin_runner/hash.rs [hunk 5]
```diff
@@ -116,6 +120,32 @@ impl HashBuiltinRunner {
...
+                .ok_or(MemoryError::InsufficientAllocatedCells)?;
+        if vm.current_step < min_step {
+            Err(MemoryError::InsufficientAllocatedCells)
+        } else {
+            let builtin = BuiltinRunner::Hash(self);
+            let used = builtin.get_used_cells(vm)?;
+            let size = cells_per_instance
+                * safe_div(&bigint!(vm.current_step), &bigint!(ratio))
+                    .map_err(|_| MemoryError::InsufficientAllocatedCells)?;
+            Ok((used, size))
```
Comment (Oppen) [inline src/vm/runners/builtin_runner/hash.rs:146]:
```suggestion
        let ratio = self._ratio as usize;
        let cells_per_instance = self.cells_per_instance;
        let min_step = ratio * self.instances_per_component as usize;
        if vm.current_step < min_step {
            Err(MemoryError::InsufficientAllocatedCells)
        } else {
            let builtin = BuiltinRunner::Hash(self);
            let used = builtin.get_used_cells(vm)?;
            let size = cells_per_instance
                * safe_div(&bigint!(vm.current_step), &bigint!(ratio))
                    .map_err(|_| MemoryError::InsufficientAllocatedCells)?;
            Ok((used, size))
```

File: src/vm/runners/builtin_runner/range_check.rs [hunk 6]
```diff
@@ -105,6 +109,32 @@ impl RangeCheckBuiltinRunner {
...
         ("range_check", (self.base, self.stop_ptr))
     }
+
+    pub fn get_used_cells_and_allocated_size(
+        self,
+        vm: &VirtualMachine,
+    ) -> Result<(usize, BigInt), MemoryError> {
+        let ratio = self
+            .ratio
+            .to_usize()
```
Comment (Oppen) [inline src/vm/runners/builtin_runner/range_check.rs:119]: `as usize`

File: src/vm/runners/builtin_runner/range_check.rs [hunk 7]
```diff
@@ -105,6 +109,32 @@ impl RangeCheckBuiltinRunner {
...
+            .ok_or(MemoryError::InsufficientAllocatedCells)?;
+        let cells_per_instance = self._cells_per_instance;
+        let min_step = ratio
+            * self
+                .instances_per_component
+                .to_usize()
```
Comment (Oppen) [inline src/vm/runners/builtin_runner/range_check.rs:125]: `as usize`

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/runners/builtin_runner/bitwise.rs [hunk 8]
```diff
@@ -115,6 +119,32 @@ impl BitwiseBuiltinRunner {
         ("bitwise", (self.base, self.stop_ptr))
     }
 
+    pub fn get_used_cells_and_allocated_size(
+        self,
+        vm: &VirtualMachine,
```
Comment (fmoletta) [inline src/vm/runners/builtin_runner/bitwise.rs:124]: Can we receive `vm.current_step` instead of the whole `vm`?

Comment (Jrigada) [reply src/vm/runners/builtin_runner/bitwise.rs:124]: We cant at least with this implementation due to the fact that we need to retrieve the used cells with the get_used_cells

**Review by fmoletta — APPROVED** [review]
LGTM!

**Review by Oppen — APPROVED** [review]

**Pull request merged**